### ISSUE TITLE
Quest: Feature flags for upcoming Edition features

### ISSUE THREAD

**Issue by mansona** [ember-learn/guides-source#2047]
This issue is intended to be a holding place for everything related to the idea of "feature flags" to unblock writing documentation for upcoming Editions - specifically Polaris.

## Summary of the problem

As part of our last Edition (Octane) we started a [long-running branch](https://github.com/ember-learn/guides-source/pull/455) that was open for almost a year. This branch was challenging to keep up to date and we needed to constantly merge our work between main and our branch to make sure that updates were not lost over the course of that year. Learning from the challenges of maintaining that long-running branch we propose that we implement a feature-flag strategy for the up-coming Edition Polaris.

## Motivation of feature flags

If we use a feature-flag strategy we can merge any updates to the documentation straight to the `main` branch without needing to keep a long-running branch up to date. It will also mean that the production [deployment of the guides](https://guides.emberjs.com/release/) will be the same place that you will be able to see the "Polaris preview guides" by toggling a flag that says you want to see the preview guides. When Polaris becomes the released Edition then the process of releasing is to just delete the feature flag for that version of the guides 👍 

## Requirements

- We need to be able to define the feature flags that are available to toggle on a version-by version basis. i.e. before version 5.9 we don't have any Polaris documentation so there should be no UI element displayed to toggle feature flags
- we will likely want to support the idea of "sub flags" e.g. as Typescript is a Pillar of Polaris we will likely want to allow people to read our documentation with the examples in Typescript and include any descriptive blocks related to just typescript. But the toggle for typescript docs will only be available after you have first toggled the polaris docs
  - we will also probably need to write a simple linting system to make sure that sub-flag marked content is only contained inside its parent sections
- we don't want to complicate the authoring format of the markdown. The guides are written using [guidemaker](https://github.com/empress/guidemaker) which has an explicit design requirement of being easy to use for documentation writers. If we need an overly complicated structure of docs to achieve feature flagging then this project will be considered a failure
  - @IgnaceMaes has proposed that we copy Vue's docs by wrapping flagged content in a div with the right class and we can then toggle them https://github.com/vuejs/docs/blob/8a4f62706df7b63de0e9f11e8a50b2c1f4bf1aff/src/guide/introduction.md?plain=1#L32 . This is a perfect solution as long as we continue to use a markdown renderer that follows the Commonmark standard of rendering markdown that is surrounded by HTML elements 👍 
  - this needs to be structured in some way, if we use classes then we need to make sure the flag classes have the same prefix so we can lint that you're only using flags that the current version of the docs supports. We could consider using `data-flag="something"` structure so we don't use flags but this could be discussed and decided upon later
- any solution that we use must not affect SEO in any way. This means that we need to make sure that there is a default implementation that works in SSG/pre-rendering so that the description field remains correct to bots

## Process

Once there is broad agreement on this plan among the core teams we will start putting together more concrete tasks on how to make this plan happen. This issue will either become a series of tasks in a list or it will link to a project that will keep track of the work 👍

**Comment by jaredgalanis** [comment]
I do understand the plan and agree with it 👍. I also do think it represents what we discussed on 6/17.

**Comment by IgnaceMaes** [comment]
Overall this looks like a good way forward!

Some thoughts:

- Agree we need to limit the nesting of flags, as otherwise it will get too complex to manage.
- I'm a fan of the `data-flag` attribute (or similar) approach compared to using classes (but that's an implementation detail)
- I don't think TypeScript really follows the same story as other new features in the way that it doesn't replace something else, so I don't think it should be a feature flag. And as long as Ember officially supports JS and TS, we need to support both. Making it use the same authoring format is probably good though. So it's more the way of visualizing the toggle where the detail really matters. (which might be seen as implementation detail as well?)

**Comment by NullVoxPopuli** [comment]
Will feature flags be in query params -- so that we can link to specific areas with flags enabled/disabled? (I think we should!)

**Comment by kategengler** [comment]
I agree with @IgnaceMaes -- I prefer the `data-flag` attribute as it would be easy to search and clean up old areas. 

I also agree that TS examples are different than a feature flag since it would be a feature flag that lived forever.

**Comment by mansona** [comment]
@kategengler I added the examples to explain what will happen with Typescript. We probably need to not call these things "feature flags" in the code because it is implying that they will get deleted eventually. TS is a "sub-flag" in my example.

Looking at the examples does it work the way that you expect it to?

**Comment by kategengler** [comment]
I think the examples are fine but I do think you have two different concepts, even if they may share implementation details. One is for a long-lived feature of the guides to toggle between different types of examples (TS vs JS) and another is feature flags which are expected to be deleted eventually.
